### PR TITLE
annotation: enable custom listener ports through annotation

### LINKED ISSUE #159: Is it possible to open a port other than 80/443 for listening?
Looking to indicate a new listener port other than 80/443.

### PR DESCRIPTION
**Pull request by rbickel**
## Checklist
- [x] The title of the PR is clear and informative
- [x] If applicable, the changes made in the PR have proper test coverage
- [x] Issues addressed by the PR are mentioned in the description followed by `Fixes`.

## Description
#159 
Implement the workitem #159, allowing to override the listener port per ingress. The feature propose a new annotation, enabling the developer to define the listtenre port

`"appgw.ingress.kubernetes.io/override-frontend-port": "444"`

If the port is withing the App Gw authorized range (1 - 64999) this listener will be created on this specific port. If an invalid port or no port is set in the annotation, the configuration will fallback on default 80 or 443.

Note: Documentation is missing. Should I update the MD documentation in this PR ?

## Fixes
#159

### CODE REVIEW

**Comment by akshaysngupta** [comment]
@rbickel Asked Biao to look as well.

File: pkg/annotations/ingress_annotations.go [hunk 1]
```diff
@@ -45,6 +45,9 @@ const (
 	// UsePrivateIPKey defines the key to determine whether to use private ip with the ingress.
 	UsePrivateIPKey = ApplicationGatewayPrefix + "/use-private-ip"
 
+	// OverrideFrontendPortKey defines the key to define a custom fronend port
```
Comment (3quanfeng) [inline pkg/annotations/ingress_annotations.go:48]: typo: custom frontend port

File: pkg/annotations/ingress_annotations.go [hunk 2]
```diff
@@ -167,6 +170,11 @@ func UsePrivateIP(ing *v1beta1.Ingress) (bool, error) {
 	return parseBool(ing, UsePrivateIPKey)
 }
 
+// OverrideFrontendPort determines whether to use a custom Frontend port
```
Comment (3quanfeng) [inline pkg/annotations/ingress_annotations.go:173]: nit: OverrideFrontendPort provides a custom frontend port to be used in listener

File: pkg/appgw/configbuilder.go [hunk 3]
```diff
@@ -214,7 +214,7 @@ func generateListenerID(ingress *v1beta1.Ingress, rule *v1beta1.IngressRule, pro
 	if protocol == n.HTTPS {
 		frontendPort = Port(443)
 	}
-	if overridePort != nil {
+	if overridePort != nil && *overridePort > 0 && *overridePort < 65000 {
```
Comment (3quanfeng) [inline pkg/appgw/configbuilder.go:217]: How about having a V5 logging to indicate a warning such as port out of boundary when annotation is enabled?

Comment (rbickel) [reply pkg/appgw/configbuilder.go]: Added the V5 for out of bound configuration

**Review by 3quanfeng — APPROVED** [review]
Thanks for contribution.

**Comment by schlickspringer** [comment]
@akshaysngupta is your review still outstanding? 

when can we expect to get this feature released? 

many thanks

**Comment by schlickspringer** [comment]
Thanks for merging that @akshaysngupta ! Is there any date when this will be packaged for a release?

**Pull request merged**